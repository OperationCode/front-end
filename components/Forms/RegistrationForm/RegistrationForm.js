import { useState } from 'react';
import { func, number, oneOfType, shape, string, boolean } from 'prop-types';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import noop from 'lodash/noop';
import { createUser } from 'common/constants/api';
import { getServerErrorMessage } from 'common/utils/api-utils';
import { validationErrorMessages } from 'common/constants/messages';
import { capitalizeFirstLetter } from 'common/utils/string-utils';
import { codeOfConduct } from 'common/constants/urls';
import Button from 'components/Buttons/Button/Button';
import Checkbox from 'components/Form/Checkbox/Checkbox';
import Form from 'components/Form/Form';
import Input from 'components/Form/Input/Input';
import Alert from 'components/Alert/Alert';
import Link from 'next/link';
import OutboundLink from 'components/OutboundLink/OutboundLink';
import styles from './RegistrationForm.module.css';

export const PASSWORD_FOR_EVERYBODY = 'nOnEeDfOrP@ssw0rd!';

const defaultValues = {
  email: '',
  'confirm-email': '',
  password: PASSWORD_FOR_EVERYBODY,
  'confirm-password': PASSWORD_FOR_EVERYBODY,
  firstName: '',
  lastName: '',
  zipcode: '',
  codeOfConduct: false,
};

/**
 * Zipcode issues solved via a trim check from Yup.
 *
 * This may seem counter-intuitive, but it's difficult to get a zipcode regex correctly.
 * See https://stackoverflow.com/questions/578406/what-is-the-ultimate-postal-code-and-zip-regex
 * for more info
 *
 */
const registrationSchema = Yup.object().shape({
  email: Yup.string()
    .required(validationErrorMessages.required)
    .email(validationErrorMessages.email),
  'confirm-email': Yup.string()
    .required(validationErrorMessages.required)
    .oneOf([Yup.ref('email')], validationErrorMessages.emailsMatch),
  password: Yup.string().required(validationErrorMessages.required),
  'confirm-password': Yup.string()
    .required(validationErrorMessages.required)
    .oneOf([Yup.ref('password')], validationErrorMessages.passwordsMatch),
  firstName: Yup.string().trim().required(validationErrorMessages.required),
  lastName: Yup.string().trim().required(validationErrorMessages.required),
  zipcode: Yup.string().trim().required(validationErrorMessages.required),
  codeOfConduct: Yup.boolean().oneOf([true], validationErrorMessages.codeOfConduct),
});

RegistrationForm.propTypes = {
  onSubmit: func,
  onSuccess: func.isRequired,
  initialValues: shape({
    email: string,
    'confirm-email': string,
    password: string,
    'confirm-password': string,
    firstName: string,
    lastName: string,
    zipcode: oneOfType([string, number]),
    codeOfConduct: boolean,
  }),
};

RegistrationForm.defaultProps = {
  onSubmit: noop,
  initialValues: defaultValues,
};

function RegistrationForm({ initialValues, onSubmit, onSuccess }) {
  const [errorMessage, setErrorMessage] = useState('');

  const handleSubmit = async (values, actions) => {
    try {
      onSubmit();
      const { token } = await createUser(values);
      await onSuccess({ user: values, token });
      actions.setSubmitting(false);
      actions.resetForm({ values: defaultValues });
    } catch (error) {
      actions.setSubmitting(false);

      const { data } = error.response;
      if (data) {
        // TODO: Create back-end ticket for checking if email has been taken for a debounced,
        // client-side validation of emails instead of waiting for submission.
        const newErrorMessage = Object.keys(data)
          .map(key => {
            const fieldName = capitalizeFirstLetter(key);

            // example: Email has already been taken.
            return `${fieldName}: ${data[key][0]}.`;
          })
          .join('\n');

        setErrorMessage(newErrorMessage);
      } else {
        setErrorMessage(getServerErrorMessage(error));
      }
    }
  };

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={handleSubmit}
      validationSchema={registrationSchema}
    >
      {({ isSubmitting }) => (
        <Form className={styles.RegistrationForm}>
          <p>
            We work closely with military veterans, service members, and spouses who are passionate
            about transitioning into the tech industry. We work with over 7,000 members who are all
            working towards relevant goals on Slack and in-person meet-ups. Membership is free!
            Unfamiliar with Slack?{` `}
            <Link href="/slack_guide">
              <a>Learn how to use it!</a>
            </Link>
          </p>

          <div className={styles.formContainer}>
            <Field
              type="email"
              name="email"
              label="Email*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="username email"
            />

            <Field
              type="email"
              name="confirm-email"
              label="Confirm Email*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="username email"
            />

            <Field
              className={styles.invisible}
              type="password"
              name="password"
              label="Password*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            <Field
              className={styles.invisible}
              type="password"
              name="confirm-password"
              label="Confirm Password*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            <Field
              type="text"
              name="firstName"
              label="First Name*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="given-name"
            />

            <Field
              type="text"
              name="lastName"
              label="Last Name*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="family-name"
            />

            <Field
              type="text"
              name="zipcode"
              label="Zipcode*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="postal-code"
            />

            <Field
              type="checkbox"
              name="codeOfConduct"
              label={
                <span>
                  I have read and agree to&nbsp;
                  <OutboundLink
                    hasIcon={false}
                    href={codeOfConduct}
                    analyticsEventLabel="Registration CoC Checkbox Link"
                  >
                    Operation Code&apos;s Code of Conduct.
                  </OutboundLink>
                  *
                </span>
              }
              component={Checkbox}
              disabled={isSubmitting}
            />
          </div>

          {errorMessage && <Alert type="error">{errorMessage}</Alert>}

          <hr className={styles.seperator} />

          <p className={styles.aside}>
            The demographic information you provide, helps us understand our community needs, ensure
            diversity, and provide specific resources to reach our mission. Thank you in advance for
            providing honest answers.
            <br />
            <span className={styles.bold}>We do not sell your information to anyone.</span>
          </p>

          <hr className={styles.seperator} />

          <Button
            className={styles.topMargin}
            type="submit"
            theme="secondary"
            disabled={isSubmitting}
          >
            Submit
          </Button>
        </Form>
      )}
    </Formik>
  );
}

export default RegistrationForm;
