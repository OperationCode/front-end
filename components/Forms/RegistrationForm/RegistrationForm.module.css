.RegistrationForm {
  width: 100%;
  display: flex;
  align-items: center;
  flex-direction: column;
}

.topMargin {
  margin-top: 1rem;
}

.RegistrationForm > p {
  margin: 1.25rem auto;
}
