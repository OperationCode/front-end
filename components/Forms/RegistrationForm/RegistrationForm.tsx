import { useState } from 'react';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import { getServerErrorMessage } from 'common/utils/api-utils';
import { validationErrorMessages } from 'common/constants/messages';
import { codeOfConduct, slackGuidelines } from 'common/constants/urls';
import Button from 'components/Buttons/Button/Button';
import Checkbox from 'components/Form/Checkbox/Checkbox';
import Form from 'components/Form/Form';
import Input from 'components/Form/Input/Input';
import Alert from 'components/Alert/Alert';
import Link from 'next/link';
import OutboundLink from 'components/OutboundLink/OutboundLink';
import axios, { AxiosError } from 'axios';

export interface RegistrationFormValues {
  email: string;
  'confirm-email': string;
  firstName: string;
  lastName: string;
  zipcode: string | number;
  codeOfConduct: boolean;
  slackGuidelines: boolean;
}

const defaultValues: RegistrationFormValues = {
  email: '',
  'confirm-email': '',
  firstName: '',
  lastName: '',
  zipcode: '',
  codeOfConduct: false,
  slackGuidelines: false,
};

/**
 * Zipcode issues solved via a trim check from Yup.
 *
 * This may seem counter-intuitive, but it's difficult to get a zipcode regex correctly.
 * See https://stackoverflow.com/questions/578406/what-is-the-ultimate-postal-code-and-zip-regex
 * for more info
 *
 */
const registrationSchema = Yup.object().shape({
  email: Yup.string()
    .required(validationErrorMessages.required)
    .email(validationErrorMessages.email),
  'confirm-email': Yup.string()
    .required(validationErrorMessages.required)
    .oneOf([Yup.ref('email')], validationErrorMessages.emailsMatch),
  firstName: Yup.string().trim().required(validationErrorMessages.required),
  lastName: Yup.string().trim().required(validationErrorMessages.required),
  zipcode: Yup.string().trim().required(validationErrorMessages.required),
  codeOfConduct: Yup.boolean().oneOf([true], validationErrorMessages.codeOfConduct),
  slackGuidelines: Yup.boolean().oneOf([true], validationErrorMessages.slackGuidelines),
});

interface RegistrationFormProps {
  initialValues?: RegistrationFormValues;
  onSubmit?: () => void;
  onSuccess: () => void;
}

export function RegistrationForm({
  initialValues = defaultValues,
  onSubmit,
  onSuccess,
}: RegistrationFormProps) {
  const [errorMessage, setErrorMessage] = useState('');

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={async (values, actions) => {
        try {
          onSubmit?.();
          await axios.post('/api/registration/new', values);
          onSuccess();
        } catch (_error: unknown) {
          const axiosError = (_error as AxiosError).response?.data as
            | Record<string, string>
            | undefined;

          const axiosErrorMessage = axiosError?.message as string | undefined;

          if (axiosErrorMessage) {
            setErrorMessage(axiosErrorMessage);
          } else {
            setErrorMessage(getServerErrorMessage(_error));
          }
        } finally {
          actions.setSubmitting(false);
        }
      }}
      validationSchema={registrationSchema}
    >
      {({ isSubmitting }) => (
        <Form className="flex flex-col items-center w-full">
          <div className="max-w-xl px-4">
            <p className="pb-4">
              We work closely with military veterans, service members, and spouses who are
              passionate about transitioning into the tech industry. We work with over 7,000 members
              who are all working towards relevant goals on Slack and in-person meet-ups. Membership
              is free! Unfamiliar with Slack?{` `}
              <Link href="/slack_guide">
                <a>Learn how to use it!</a>
              </Link>
            </p>
          </div>

          <div className="flex flex-col justify-between max-w-md">
            <Field
              type="email"
              name="email"
              label="Email*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="username email"
            />

            <Field
              type="email"
              name="confirm-email"
              label="Confirm Email*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="username email"
            />

            <Field
              className="hidden"
              type="password"
              name="password"
              label="Password*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            <Field
              className="hidden"
              type="password"
              name="confirm-password"
              label="Confirm Password*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            <Field
              type="text"
              name="firstName"
              label="First Name*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="given-name"
            />

            <Field
              type="text"
              name="lastName"
              label="Last Name*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="family-name"
            />

            <Field
              type="text"
              name="zipcode"
              label="Zipcode*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="postal-code"
            />

            <Field
              type="checkbox"
              name="codeOfConduct"
              label={
                <span>
                  I have read and agree to&nbsp;
                  <OutboundLink
                    hasIcon
                    href={codeOfConduct}
                    analyticsEventLabel="Registration CoC Checkbox Link"
                  >
                    Operation Code&apos;s Code of Conduct.
                  </OutboundLink>
                  *
                </span>
              }
              component={Checkbox}
              disabled={isSubmitting}
            />

            <Field
              type="checkbox"
              name="slackGuidelines"
              label={
                <span>
                  I have read and agree to&nbsp;
                  <OutboundLink
                    hasIcon
                    href={slackGuidelines}
                    analyticsEventLabel="Registration Slack Guidelines Checkbox Link"
                  >
                    Operation Code&apos;s Slack Community Guidelines.
                  </OutboundLink>
                  *
                </span>
              }
              component={Checkbox}
              disabled={isSubmitting}
            />
          </div>

          {errorMessage && <Alert type="error">{errorMessage}</Alert>}

          <div className="max-w-md px-3 mt-5">
            <Button type="submit" theme="secondary" disabled={isSubmitting}>
              Submit
            </Button>
            <p className="pt-5 text-xs">
              The demographic information you provide, helps us understand our community needs,
              ensure diversity, and provide specific resources to reach our mission. Thank you in
              advance for providing honest answers.&nbsp;
              <span className="font-bold">We do not sell your information to anyone.</span>
            </p>
          </div>
        </Form>
      )}
    </Formik>
  );
}
