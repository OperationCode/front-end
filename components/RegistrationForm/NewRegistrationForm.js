// import React from 'react';
// import PropTypes from 'prop-types';
// import { Formik, Field, ErrorMessage } from 'formik';
// import * as Yup from 'yup';
// import Button from 'components/Button/Button';

// class MultiPageForm extends React.Component {
//   static propTypes = {
//     children: PropTypes.node.isRequired,
//     initialValues: PropTypes.shape({
//       email: PropTypes.string,
//       'confirm-email': PropTypes.string,
//       password: PropTypes.string,
//       'confirm-password': PropTypes.string,
//       firstName: PropTypes.string,
//       lastName: PropTypes.string,
//       zipcode: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
//     }),
//     onSubmit: PropTypes.func.isRequired,
//     startingPageNumber: PropTypes.number,
//     validationSchema: PropTypes.oneOf(Yup.object().shape()).isRequired,
//   };

//   static defaultProps = {
//     initialValues: {
//       email: '',
//       'confirm-email': '',
//       password: '',
//       'confirm-password': '',
//       firstName: '',
//       lastName: '',
//       zipcode: '',
//     },
//     startingPageNumber: 0,
//   };

//   static Page = ({ children }) => children;

//   state = {
//     /* eslint-disable react/destructuring-assignment */
//     pageNumber: this.props.startingPageNumber,
//     values: this.props.initialValues,
//     /* eslint-enable react/destructuring-assignment */
//   };

//   goToNextPage = values => {
//     this.setState(previousState => ({
//       pageNumber: previousState.pageNumber + 1,
//       values,
//     }));
//   };

//   goToPreviousPage = () => {
//     this.setState(previousState => ({
//       pageNumber: previousState.pageNumber - 1,
//     }));
//   };

//   validateCurrentPage = values => {
//     const { props, state } = this;
//     const activePage = React.Children.toArray(props.children)[state.pageNumber];
//     return activePage.props.validate ? activePage.props.validate(values) : {};
//   };

//   handleSubmit = (values, formikBag) => {
//     const { children, onSubmit } = this.props;
//     const { pageNumber } = this.state;

//     const isLastPage = pageNumber === React.Children.count(children) - 1;
//     if (isLastPage) {
//       return onSubmit(values, formikBag);
//     }

//     formikBag.setTouched({});
//     formikBag.setSubmitting(false);
//     this.goToNextPage(values);
//   };

//   render() {
//     const { props, state } = this;

//     const activePage = React.Children.toArray(props.children)[state.pageNumber];
//     const isLastPage = state.pageNumber === React.Children.count(props.children) - 1;

//     return (
//       <Formik
//         initialValues={state.values}
//         validate={this.validateCurrentPage}
//         onSubmit={this.handleSubmit}
//         render={({ handleSubmit, isSubmitting }) => (
//           <form onSubmit={handleSubmit}>
//             {activePage}

//             <div className="buttons">
//               {state.pageNumber > 0 && (
//              <Button theme="secondary" disabled={isSubmitting} onClick={this.goToPreviousPage}>
//                   « Previous
//                 </Button>
//               )}

//               {isLastPage ? (
//                 <Button type="submit" theme="secondary" disabled={isSubmitting}>
//                   Submit
//                 </Button>
//               ) : (
//                 <Button type="submit" theme="secondary" disabled={isSubmitting}>
//                   Next »
//                 </Button>
//               )}
//             </div>
//           </form>
//         )}
//       />
//     );
//   }
// }

// const App = () => (
//   <div className="App">
//     <MultiPageForm
//       initialValues={{
//         firstName: '',
//         lastName: '',
//         email: '',
//         favoriteColor: '',
//       }}
//       onSubmit={(values, actions) => {
//         window.alert(JSON.stringify(values, null, 2)); // eslint-disable-line no-alert
//         actions.setSubmitting(false);
//       }}
//     >
//       <MultiPageForm.Page>
//         <div>
//           <label>First Name</label>
//           <Field
//             name="firstName"
//             component="input"
//             type="text"
//             placeholder="First Name"
//             validate={required}
//           />
//           <ErrorMessage name="firstName" component="div" className="field-error" />
//         </div>
//         <div>
//           <label>Last Name</label>
//           <Field
//             name="lastName"
//             component="input"
//             type="text"
//             placeholder="Last Name"
//             validate={required}
//           />
//           <ErrorMessage name="lastName" component="div" className="field-error" />
//         </div>
//       </MultiPageForm.Page>
//       <MultiPageForm.Page
//         validate={values => {
//           const errors = {};
//           if (!values.email) {
//             errors.email = 'Required';
//           }
//           if (!values.favoriteColor) {
//             errors.favoriteColor = 'Required';
//           }
//           return errors;
//         }}
//       >
//         <div>
//           <label>Email</label>
//           <Field name="email" component="input" type="email" placeholder="Email" />
//           <ErrorMessage name="email" component="div" className="field-error" />
//         </div>
//         <div>
//           <label>Favorite Color</label>
//           <Field name="favoriteColor" component="select">
//             <option value="">Select a Color</option>
//             <option value="#ff0000">❤️ Red</option>
//             <option value="#00ff00">💚 Green</option>
//             <option value="#0000ff">💙 Blue</option>
//           </Field>
//           <ErrorMessage name="favoriteColor" component="div" className="field-error" />
//         </div>
//       </MultiPageForm.Page>
//     </MultiPageForm>
//   </div>
// );

// export default App;
