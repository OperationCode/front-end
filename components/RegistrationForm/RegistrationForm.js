import React, { Component } from 'react';
import { func, number, oneOfType, shape, string } from 'prop-types';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import { createUser } from 'common/constants/api';
import { getErrorMessage } from 'common/utils/api-utils';
import { validationErrorMessages } from 'common/constants/messages';
import { minimumPasswordLength } from 'common/constants/validations';
import { capitalizeFirstLetter } from 'common/utils/string-utils';
import { isMinPasswordStrength, isValidZipcode } from 'common/utils/validator-utils';
import Button from 'components/Button/Button';
import Form from 'components/Form/Form';
import Input from 'components/Form/Input/Input';
import Alert from 'components/Alert/Alert';
import styles from './RegistrationForm.css';

const registrationSchema = Yup.object().shape({
  email: Yup.string()
    .required(validationErrorMessages.required)
    .email(validationErrorMessages.email),
  'confirm-email': Yup.string()
    .required(validationErrorMessages.required)
    .oneOf([Yup.ref('email')], validationErrorMessages.emailMatch),
  password: Yup.string()
    .required(validationErrorMessages.required)
    .min(minimumPasswordLength, validationErrorMessages.length(minimumPasswordLength))
    .test('password-strength', validationErrorMessages.password, isMinPasswordStrength),
  'confirm-password': Yup.string()
    .required(validationErrorMessages.required)
    .oneOf([Yup.ref('password')], validationErrorMessages.passwordMatch),
  firstName: Yup.string().required(validationErrorMessages.required),
  lastName: Yup.string().required(validationErrorMessages.required),
  zipcode: Yup.string()
    .required(validationErrorMessages.required)
    .test('zipcode', validationErrorMessages.zipcode, isValidZipcode),
});

class RegistrationForm extends Component {
  static propTypes = {
    onSuccess: func.isRequired,
    initialValues: shape({
      email: string,
      'confirm-email': string,
      password: string,
      'confirm-password': string,
      firstName: string,
      lastName: string,
      zipcode: oneOfType([string, number]),
    }),
  };

  static defaultProps = {
    initialValues: {
      email: '',
      'confirm-email': '',
      password: '',
      'confirm-password': '',
      firstName: '',
      lastName: '',
      zipcode: '',
    },
  };

  state = {
    errorMessage: '',
  };

  handleSubmit = async (values, actions) => {
    const { onSuccess } = this.props;

    try {
      const { token } = await createUser(values);
      actions.setSubmitting(false);
      actions.resetForm();

      await onSuccess({ user: { ...values, slackName: '', isMentor: false }, token });
    } catch (error) {
      actions.setSubmitting(false);

      const { data } = error.response;

      if (data) {
        // TODO: Create back-end ticket for checking if email has been taken for a debounced,
        // client-side validation of emails instead of waiting for submission.
        const errorMessage = Object.keys(data)
          .map(key => {
            const fieldName = capitalizeFirstLetter(key);

            // example: Email has already been taken.
            return `${fieldName} ${data[key][0]}.`;
          })
          .join('\n');

        this.setState({ errorMessage });
      } else {
        this.setState({ errorMessage: getErrorMessage(error) });
      }
    }
  };

  render() {
    const { props, state } = this;

    return (
      <Formik
        initialValues={props.initialValues}
        onSubmit={this.handleSubmit}
        validationSchema={registrationSchema}
      >
        {({ isSubmitting }) => (
          <Form className={styles.RegistrationForm}>
            <div className={styles.row}>
              <Field
                type="email"
                name="email"
                label="Email*"
                component={Input}
                disabled={isSubmitting}
                autoComplete="username email"
              />

              <Field
                type="email"
                name="confirm-email"
                label="Confirm Email*"
                component={Input}
                disabled={isSubmitting}
                autoComplete="username email"
              />
            </div>

            <div className={styles.row}>
              <Field
                type="password"
                name="password"
                label="Password*"
                component={Input}
                disabled={isSubmitting}
                autoComplete="new-password"
              />

              <Field
                type="password"
                name="confirm-password"
                label="Confirm Password*"
                component={Input}
                disabled={isSubmitting}
                autoComplete="new-password"
              />
            </div>

            <div className={styles.row}>
              <Field
                type="text"
                name="firstName"
                label="First Name*"
                component={Input}
                disabled={isSubmitting}
                autoComplete="given-name"
              />

              <Field
                type="text"
                name="lastName"
                label="Last Name*"
                component={Input}
                disabled={isSubmitting}
                autoComplete="family-name"
              />
            </div>

            <div className={styles.row}>
              <Field
                type="text"
                name="zipcode"
                label="Zipcode*"
                component={Input}
                disabled={isSubmitting}
                autoComplete="postal-code"
              />
            </div>

            <div className={styles.row}>
              <Alert isOpen={Boolean(state.errorMessage)} type="error">
                {state.errorMessage}
              </Alert>
            </div>

            <div className={styles.row}>
              <Button
                className={styles.topMargin}
                type="submit"
                theme="secondary"
                disabled={isSubmitting}
              >
                Submit
              </Button>
            </div>
          </Form>
        )}
      </Formik>
    );
  }
}

export default RegistrationForm;
