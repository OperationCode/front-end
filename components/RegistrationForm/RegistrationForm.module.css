.RegistrationForm {
  width: 100%;
  display: flex;
  align-items: center;
  flex-direction: column;
}

.topMargin {
  margin-top: 1rem;
}

.RegistrationForm > p {
  margin: 1.25rem auto;
}

<<<<<<< HEAD
.RegistrationForm label {
  display: block;
  margin: 1rem;
}
=======
.Field > .checkWidth {
  min-width: 0px !important;
  width: 40px;
  height: 40px;
}
>>>>>>> 98f86fed... stylint
