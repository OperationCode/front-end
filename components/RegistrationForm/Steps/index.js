import EmailPasswordStep from './EmailPasswordStep/EmailPasswordStep';
import MilitaryStep from './MilitaryStep/MilitaryStep';
import PersonaStep from './PersonaStep/PersonaStep';
import TechnologyStep from './TechnologyStep/TechnologyStep';
import UserDetailsStep from './UserDetailsStep/UserDetailsStep';

export default {
  EmailPasswordStep,
  MilitaryStep,
  PersonaStep,
  TechnologyStep,
  UserDetailsStep,
};
