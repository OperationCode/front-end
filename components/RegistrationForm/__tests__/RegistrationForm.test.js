import React from 'react';
import { mount } from 'enzyme';
import { networkErrorMessages, validationErrorMessages } from 'common/constants/messages';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import mockUser from 'test-utils/mockGenerators/mockUser';
import mockPassword from 'test-utils/mockGenerators/mockPassword';
import OperationCodeAPIMock from 'test-utils/mocks/apiMock';
import asyncRenderDiff from 'test-utils/asyncRenderDiff';
import wait from 'test-utils/wait';
import RegistrationForm from '../RegistrationForm';

afterEach(() => {
  OperationCodeAPIMock.reset();
});

describe('RegistrationForm', () => {
  it('should render with required props', () => {
    createSnapshotTest(<RegistrationForm onSuccess={jest.fn()} />);
  });

  it('should display required error message when blurring past email input', async () => {
    const wrapper = mount(<RegistrationForm onSuccess={jest.fn()} />);

    wrapper.find('input#email').simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="email"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.required);
  });

  it('should show error when providing non-email to email input', async () => {
    const wrapper = mount(<RegistrationForm onSuccess={jest.fn()} />);

    wrapper
      .find('input#email')
      .simulate('change', { target: { id: 'email', value: 'email' } })
      .simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="email"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.email);
  });

  it('should show "password required" message when blurring past input', async () => {
    const wrapper = mount(<RegistrationForm onSuccess={jest.fn()} />);

    wrapper.find('input#password').simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="password"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.required);
  });

  it('should show "invalid password" message when focusing off an invalid password', async () => {
    const wrapper = mount(<RegistrationForm onSuccess={jest.fn()} />);

    const stringWithoutNumber = 'SillyPassword';

    wrapper
      .find('input#password')
      .simulate('change', { target: { id: 'password', value: stringWithoutNumber } })
      .simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="password"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.password);
  });

  it('should display password match message when both password inputs do not match', async () => {
    const wrapper = mount(<RegistrationForm onSuccess={jest.fn()} />);

    wrapper
      .find('input#password')
      .simulate('change', {
        target: { id: 'password', value: mockPassword() },
      })
      .simulate('blur');

    wrapper
      .find('input#confirm-password')
      .simulate('change', {
        target: { id: 'confirm-password', value: 'something-else' },
      })
      .simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('input#confirm-password')
        .closest('Input')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.passwordMatch);
  });

  it('should submit with valid data in form', async () => {
    const user = mockUser();

    OperationCodeAPIMock.onPost('users', { user }).reply(200, {
      token: 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9',
    });

    const successSpy = jest.fn();
    const wrapper = mount(<RegistrationForm onSuccess={successSpy} initialState={user} />);

    wrapper.find('Button').simulate('submit');
    await asyncRenderDiff(wrapper);

    await wait(() => {
      expect(successSpy).toHaveBeenCalled();
      expect(OperationCodeAPIMock.history.post.length).toBeGreaterThan(0);
    });
  });

  it('should NOT submit with some invalid data in form', async () => {
    const successSpy = jest.fn();

    const invalidFormValues = {
      email: 'not-an-email',
      'confirm-email': 'not-the-same',
      password: 'notvalid',
      'confirm-password': 'notmatching',
      firstName: '',
      lastName: '',
      zipcode: '',
    };

    const wrapper = mount(
      <RegistrationForm onSuccess={successSpy} initialValues={invalidFormValues} />,
    );

    wrapper.find('Button').simulate('submit');
    await asyncRenderDiff(wrapper);

    await wait(() => {
      expect(successSpy).not.toHaveBeenCalled();
      expect(OperationCodeAPIMock.history.post.length).not.toBeGreaterThan(0);
    });

    // All fields + 1 because of always-rendered form error (as opposed to all the field errors)
    expect(wrapper.find('Alert').children()).toHaveLength(
      Object.keys(invalidFormValues).length + 1,
    );
  });

  it('should show "email already registered" message for dupe email registration', async () => {
    const existingUser = mockUser('kylemh.email12@gmail.com');

    OperationCodeAPIMock.onPost('users', {
      user: {
        email: existingUser.email,
        password: existingUser.password,
        first_name: existingUser.firstName,
        last_name: existingUser.lastName,
        zip: existingUser.zipcode,
      },
    }).reply(422, { email: ['has been taken'] });

    const successSpy = jest.fn();
    const wrapper = mount(<RegistrationForm onSuccess={successSpy} initialValues={existingUser} />);

    wrapper.find('Button').simulate('submit');
    await asyncRenderDiff(wrapper);

    expect(successSpy).not.toHaveBeenCalled();
    expect(
      wrapper
        .find('Alert')
        .last()
        .text(),
    ).toStrictEqual('Email has been taken.');
  });

  it('should show a helpful error if the server is down', async () => {
    const user = mockUser();

    OperationCodeAPIMock.onPost('users', {
      user: {
        email: user.email,
        password: user.password,
        first_name: user.firstName,
        last_name: user.lastName,
        zip: user.zipcode,
      },
    }).reply(503);

    const successSpy = jest.fn();
    const wrapper = mount(<RegistrationForm onSuccess={successSpy} initialValues={user} />);

    wrapper.find('button[type="submit"]').simulate('submit');
    await asyncRenderDiff(wrapper);

    expect(successSpy).not.toHaveBeenCalled();
    expect(
      wrapper
        .find('Alert')
        .last()
        .text(),
    ).toStrictEqual(networkErrorMessages.serverDown);
  });
});
