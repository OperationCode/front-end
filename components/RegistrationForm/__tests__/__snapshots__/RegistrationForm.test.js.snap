// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm should render with required props 1`] = `
<form
  className="RegistrationForm"
  noValidate={true}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="email"
      >
        Email*
      </label>
      <div
        className="inputFeedbackGrouping"
      >
        <input
          autoComplete="username email"
          className="Input"
          disabled={false}
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="confirm-email"
      >
        Confirm Email*
      </label>
      <div
        className="inputFeedbackGrouping"
      >
        <input
          autoComplete="username email"
          className="Input"
          disabled={false}
          id="confirm-email"
          name="confirm-email"
          onBlur={[Function]}
          onChange={[Function]}
          type="email"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="password"
      >
        Password*
      </label>
      <div
        className="inputFeedbackGrouping"
      >
        <input
          autoComplete="new-password"
          className="Input"
          disabled={false}
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="confirm-password"
      >
        Confirm Password*
      </label>
      <div
        className="inputFeedbackGrouping"
      >
        <input
          autoComplete="new-password"
          className="Input"
          disabled={false}
          id="confirm-password"
          name="confirm-password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="firstName"
      >
        First Name*
      </label>
      <div
        className="inputFeedbackGrouping"
      >
        <input
          autoComplete="given-name"
          className="Input"
          disabled={false}
          id="firstName"
          name="firstName"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="lastName"
      >
        Last Name*
      </label>
      <div
        className="inputFeedbackGrouping"
      >
        <input
          autoComplete="family-name"
          className="Input"
          disabled={false}
          id="lastName"
          name="lastName"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="field"
    >
      <label
        className="Label"
        htmlFor="zipcode"
      >
        Zipcode*
      </label>
      <div
        className="inputFeedbackGrouping"
      >
        <input
          autoComplete="postal-code"
          className="Input"
          disabled={false}
          id="zipcode"
          name="zipcode"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="Alert error"
      role="alert"
    >
      
    </div>
  </div>
  <div
    className="row"
  >
    <button
      className="Button topMargin secondary"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
